File: 20-Matrix-Exponentiation/java/MatrixExpFib.java
public class MatrixExpFib { static final long MOD=(long)1e9+7;
    static long[][] multiply(long[][] A,long[][] B){int n=A.length;long[][] C=new long[n][n];
    for(int i=0;i<n;i++)for(int k=0;k<n;k++)for(int j=0;j<n;j++)C[i][j]=(C[i][j]+A[i][k]*B[k][j])%MOD;return C;}
    static long[][] matPow(long[][] A,long p){int n=A.length;long[][] R=new long[n][n];for(int i=0;i<n;i++)R[i][i]=1;
    while(p>0){if((p&1)==1)R=multiply(R,A);A=multiply(A,A);p>>=1;}return R;}
    static long fib(long n){if(n<=1)return n;long[][] M={{1,1},{1,0}};return matPow(M,n-1)[0][0];}
    public static void main(String[] args){System.out.println("fib(10)="+fib(10));System.out.println("fib(50)="+fib(50));}}

File: 20-Matrix-Exponentiation/cpp/matrix_exp_fib.cpp
/* Fibonacci using Matrix Exponentiation | Time: O(log n) */
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<vector<ll>> Matrix;
const ll MOD=1e9+7;
Matrix multiply(Matrix& A,Matrix& B){int n=A.size();Matrix C(n,vector<ll>(n,0));
for(int i=0;i<n;i++)for(int k=0;k<n;k++)for(int j=0;j<n;j++)C[i][j]=(C[i][j]+A[i][k]*B[k][j])%MOD;return C;}
Matrix matPow(Matrix A,ll p){int n=A.size();Matrix R(n,vector<ll>(n,0));for(int i=0;i<n;i++)R[i][i]=1;
while(p>0){if(p&1)R=multiply(R,A);A=multiply(A,A);p>>=1;}return R;}
ll fib(ll n){if(n<=1)return n;Matrix M={{1,1},{1,0}};auto R=matPow(M,n-1);return R[0][0];}
int main(){cout<<"fib(10)="<<fib(10)<<endl;cout<<"fib(50)="<<fib(50)<<endl;}
